### PR TITLE
feat(txpool): Pending and Queued Transactions

### PR DESCRIPTION
**Pull request by calbera**
using the Cosmos SDK Priority Nonce Mempool

### CODE REVIEW

File: cosmos/x/evm/plugins/txpool/mempool/mempool.go [hunk 1]
```diff
@@ -114,20 +114,65 @@ func (etp *EthTxPool) Get(hash common.Hash) *coretypes.Transaction {
...
 
-// GetTransactions is called when the mempool is retrieved.
+// Pending is called when txs in the mempool are retrieved.
 func (etp *EthTxPool) Pending(bool) map[common.Address]coretypes.Transactions {
 	etp.mu.RLock()
 	defer etp.mu.RUnlock()
 
-	// TODO: implement me.
-	return nil
+	allNonces := etp.PriorityNonceMempool.senderIndices
```
[the remark this answers is not in the document]
Comment (calbera) [reply cosmos/x/evm/plugins/txpool/mempool/mempool.go:122]: I think the obvious optimization is to keep track of pending/queued lists separately. 1) all operations are efficient 2) no specific implementation required.

File: cosmos/crypto/keys/ethsecp256k1/keys.go [hunk 2]
```diff
@@ -51,6 +52,24 @@ const (
 // Compile-time type assertion.
 var _ cryptotypes.PrivKey = &PrivKey{}
 
+// PubkeyFromTx returns the public key of the signer of the transaction.
+func PubkeyFromTx(signedTx *coretypes.Transaction, signer coretypes.Signer) (*PubKey, error) {
```
Comment (itsdevbear) [inline cosmos/crypto/keys/ethsecp256k1/keys.go:56]: Remove this function (maybe just need to merge main)

File: cosmos/x/evm/plugins/txpool/mempool/mempool.go [hunk 3]
```diff
@@ -22,89 +22,192 @@ package mempool
...
-	if err := etp.Mempool.Insert(ctx, tx); err != nil {
+	if err := etp.PriorityNonceMempool.Insert(ctx, tx); err != nil {
 		return err
 	}
 
 	// We want to cache
-	etr, ok := utils.GetAs[*types.EthTransactionRequest](tx.GetMsgs()[0])
-	if !ok {
-		return nil
+	if ethTx := GetAsEthTx(tx); ethTx != nil {
```
Comment (itsdevbear) [inline cosmos/x/evm/plugins/txpool/mempool/mempool.go:79]: GetAsEthTx maybe should go in evm/types?

**Review by itsdevbear — APPROVED** [review]
Minor things then good to merge. 


TODO: integration tests, and optimization (we both ack it's not the most efficient)

File: cosmos/x/evm/plugins/txpool/mempool/mempool.go (hunk 3 above)
Comment (itsdevbear) [reply cosmos/x/evm/plugins/txpool/mempool/mempool.go:79]: Actually you can cast the sdk.Tx to EthTransactionRequest and then use .AsTransaction()

File: cosmos/x/evm/types/tx.go [hunk 4]
```diff
@@ -132,3 +133,12 @@ func (etr *EthTransactionRequest) ValidateBasic() error {
 
 	return nil
 }
+
+// GetAsEthTx is a helper function to get an EthTx from a sdk.Tx.
+func GetAsEthTx(tx sdk.Tx) *coretypes.Transaction {
```
Comment (itsdevbear) [inline cosmos/x/evm/types/tx.go:138]: Just use the one on the transaction tbh.

Comment (calbera) [reply cosmos/x/evm/types/tx.go:138]: How else do you do it? The sdk.Tx only has `GetMsgs()` on it. One of te messages is EthTransactionRequest

Comment (itsdevbear) [reply cosmos/x/evm/types/tx.go:138]: oh I see what you did yeah makes sense.

**Pull request merged**